extension/src/service/klarna-service.js

```diff
@@ -0,0 +1,86 @@
+const {
+  ADYEN_PERCENTAGE_MINOR_UNIT,
+  DEFAULT_PAYMENT_LANGUAGE,
+  KLARNA_DEFAULT_LINE_ITEM_NAME
+} = require('../config/klarna-constants')
+
+function createLineItems (payment, cart) {
+  const lineItems = []
+  const locales = _getLocales(cart, payment)
+
+  cart.lineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromLineItem(item, locales))
+  })
+
+  cart.customLineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromCustomLineItem(item, locales))
+  })
+
+  lineItems.push(_createShippingInfoAdyenLineItem(cart.shippingInfo))
+
+  return lineItems
+}
+
+function _getLocales (cart, payment) {
+  const locales = []
+  let paymentLanguage = payment.custom && payment.custom.fields['languageCode']
+  if (!paymentLanguage)
+    paymentLanguage = cart.locale
+  if (paymentLanguage)
+    locales.push(paymentLanguage)
+  if (!paymentLanguage || paymentLanguage !== DEFAULT_PAYMENT_LANGUAGE)
+    locales.push(DEFAULT_PAYMENT_LANGUAGE)
+
+  return locales
+}
+
+function _createAdyenLineItemFromLineItem (ctpLineItem, locales) {
+  return {
+    id: ctpLineItem.variant.sku,
+    quantity: ctpLineItem.quantity,
+    description: _localizeOrFallback(ctpLineItem.name, locales, KLARNA_DEFAULT_LINE_ITEM_NAME),
+    amountIncludingTax: ctpLineItem.price.value.centAmount,
+    taxPercentage: ctpLineItem.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
+  }
+}
+
+function _createAdyenLineItemFromCustomLineItem (ctpLineItem, locales) {
+  return {
+    id: ctpLineItem.id,
+    quantity: ctpLineItem.quantity,
+    description: _localizeOrFallback(ctpLineItem.name, locales, KLARNA_DEFAULT_LINE_ITEM_NAME),
+    amountIncludingTax: ctpLineItem.money.centAmount,
+    taxPercentage: ctpLineItem.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
+  }
+}
+
+function _createShippingInfoAdyenLineItem (shippingInfo) {
+  return {
+    id: `${shippingInfo.shippingMethodName}`,
+    quantity: 1,
+    description: _getShippingMethodDescription(shippingInfo),
+    amountIncludingTax: shippingInfo.price.centAmount,
+    taxPercentage: shippingInfo.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
+  }
+}
+
+function _getShippingMethodDescription (shippingInfo) {
+  const shippingMethod = shippingInfo.shippingMethod.obj
+  if (shippingMethod)
+    return shippingMethod.description
+  return shippingInfo.shippingMethodName
+}
+
+function _localizeOrFallback (localizedString,
+                              locales,
+                              fallback) {
+  let value
```

Comment:
refactor point: maybe write the code filtering code with the functional way (maybe sth like find) instead of imperative style ?

Reply:
I don't know how to refactor 🤔 Can you give your suggestion?

Reply:
Did you try this: locales.find(element => localisedString[element]) ? Even it could be better with additional map.

```
const locale = locales.find(locale => localizedString[locale])
return locale ? localizedString[locale] : fallback
```

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/294/commits/9e1dda9c44de2925f748322e60260b8b15f95fa0

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/294/commits/77a5ed9c4c2351520d5209402b927fb9b6967632